In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from keras.layers.core import Dense, Activation, Dropout
from keras.layers.recurrent import LSTM
from keras.models import Sequential
%matplotlib inline

# Display only top 5 and bottom 5 rows
pd.set_option('display.max_rows', 10)

Using TensorFlow backend.


In [2]:
# Set random seed for reproducibility
# Note: CuDNN is usually non-deterministic
# (can't determine which of the ~3000 threads finish earlier)
# And floating points reduction is not perfectly associative due to ULP rounding
import numpy as np
np.random.seed(1337)
import tensorflow as tf
tf.set_random_seed(1337)

In [3]:
df_train = pd.read_csv('./data/train_aWnotuB.csv', parse_dates=[0], infer_datetime_format=True)
df_train

DateTime  Junction  Vehicles           ID
0     2015-11-01 00:00:00         1        15  20151101001
1     2015-11-01 01:00:00         1        13  20151101011
2     2015-11-01 02:00:00         1        10  20151101021
3     2015-11-01 03:00:00         1         7  20151101031
4     2015-11-01 04:00:00         1         9  20151101041
...                   ...       ...       ...          ...
48115 2017-06-30 19:00:00         4        11  20170630194
48116 2017-06-30 20:00:00         4        30  20170630204
48117 2017-06-30 21:00:00         4        16  20170630214
48118 2017-06-30 22:00:00         4        22  20170630224
48119 2017-06-30 23:00:00         4        12  20170630234

[48120 rows x 4 columns]

In [4]:
train = df_train.pivot(index='DateTime',columns='Junction', values='Vehicles')
train

Junction                 1     2     3     4
DateTime                                    
2015-11-01 00:00:00   15.0   6.0   9.0   NaN
2015-11-01 01:00:00   13.0   6.0   7.0   NaN
2015-11-01 02:00:00   10.0   5.0   5.0   NaN
2015-11-01 03:00:00    7.0   6.0   1.0   NaN
2015-11-01 04:00:00    9.0   7.0   2.0   NaN
...                    ...   ...   ...   ...
2017-06-30 19:00:00  105.0  34.0  33.0  11.0
2017-06-30 20:00:00   96.0  35.0  31.0  30.0
2017-06-30 21:00:00   90.0  31.0  28.0  16.0
2017-06-30 22:00:00   84.0  29.0  26.0  22.0
2017-06-30 23:00:00   78.0  27.0  39.0  12.0

[14592 rows x 4 columns]

### Remove Nan (0 vehicle)

In [5]:
train = train.fillna(0)

# Generate rolling forecast features

In [6]:
nb_forecast_per_junction = 24 * (31 + 31 + 30 + 31) # Days in jul + aug + sep + oct

In [7]:
nb_forecast_per_junction

2952

In [8]:
nb_forecast_per_junction * 4

11808

That will certainly not fit in the GPU VRAM --> We will get inspiration from seq2seq models and do a sliding window of time that matches a week.

In [9]:
num_feats = 4
seq_len = 24 * 7 * 2 # Sequence of one week of traffic: 168
num_outputs = 4
num_hidden = 80 # Keep 80 * 2 weeks of hidden state
bs = 512
epochs = 50

We originally had 14592 rows, we will generate sequences of predictions from it.

In [10]:
def make_input_seqs(data, seq_len, train_split=0.9):
    seq_len = seq_len + 1
    result = []
    for index in range(len(data) - seq_len):
        result.append(data[index: index + seq_len, :])
    result = np.array(result) # shape (14423, 168, 4)
    train_ind = round(train_split * result.shape[0])
    train = result[:int(train_ind), :, :]
    x_train = train[:, :-1, :]
    y_train = train[:, -1, :]
    x_test = result[int(train_ind):, :-1, :]
    y_test = result[int(train_ind):, -1, :]

    return [x_train, y_train, x_test, y_test]

In [11]:
X_train, y_train, X_test, y_test = make_input_seqs(train.values, seq_len)

# Generate model

## Loss function

In [12]:
import keras.backend as K

def root_mean_squared_error(y_true, y_pred):
        return K.sqrt(K.mean(K.square(y_pred - y_true), axis=-1)) 

## Neural net

In [13]:
def Net(num_feats, seq_len, num_hidden, num_outputs):
    model = Sequential()

    # Encoder RNN
    model.add(LSTM(
        input_shape=(seq_len,num_feats),
        return_sequences=True, units=seq_len))
    model.add(Dropout(0.2))
    
    # Deoder RNN
    model.add(LSTM(
        num_hidden,
        return_sequences=False))
    model.add(Dropout(0.2))

    model.add(Dense(num_outputs))
    model.add(Activation("linear"))

    model.compile(loss= root_mean_squared_error, optimizer="rmsprop")
    return model

In [14]:
model = Net(num_feats, seq_len, num_hidden, num_outputs)

In [15]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_1 (LSTM)                (None, 336, 336)          458304    
_________________________________________________________________
dropout_1 (Dropout)          (None, 336, 336)          0         
_________________________________________________________________
lstm_2 (LSTM)                (None, 80)                133440    
_________________________________________________________________
dropout_2 (Dropout)          (None, 80)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 4)                 324       
_________________________________________________________________
activation_1 (Activation)    (None, 4)                 0         
Total params: 592,068
Trainable params: 592,068
Non-trainable params: 0
_________________________________________________________________


### Callbacks

In [16]:
from keras.callbacks import History, ModelCheckpoint, CSVLogger, EarlyStopping
import time

In [ ]:
history = History()
checkpointer = ModelCheckpoint(filepath="checkpoints/" + time.strftime("%Y-%m-%d_%H%M-")+"seq2seq.hdf5",
                               verbose=1, save_best_only=False)
csv_logger = CSVLogger("checkpoints/" + time.strftime("%Y-%m-%d_%H%M-")+'training.log')
early_stop = EarlyStopping(monitor='val_loss', min_delta=0, patience=2, verbose=1, mode='auto')

# Training

In [ ]:
model.fit(X_train, y_train,
          batch_size=bs,
          epochs=epochs,
          validation_split=0.05,
          callbacks=[history,checkpointer,csv_logger,early_stop])

Train on 12188 samples, validate on 642 samples
Epoch 1/50

11776/12188 [===========================>..] - ETA: 1s - loss: 17.6295Epoch 00001: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 17.5855 - val_loss: 27.9264
Epoch 2/50

11776/12188 [===========================>..] - ETA: 1s - loss: 15.5601Epoch 00002: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 15.5324 - val_loss: 26.5664
Epoch 3/50

11776/12188 [===========================>..] - ETA: 1s - loss: 14.5285Epoch 00003: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 14.4967 - val_loss: 25.5097
Epoch 4/50

11776/12188 [===========================>..] - ETA: 1s - loss: 13.6759Epoch 00004: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 13.6

11776/12188 [===========================>..] - ETA: 1s - loss: 12.7999Epoch 00005: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 12.8087 - val_loss: 23.0287
Epoch 6/50

11776/12188 [===========================>..] - ETA: 1s - loss: 12.0529Epoch 00006: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 12.0347 - val_loss: 22.0064
Epoch 7/50

11776/12188 [===========================>..] - ETA: 1s - loss: 11.3101Epoch 00007: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 11.3182 - val_loss: 21.0771
Epoch 8/50

11776/12188 [===========================>..] - ETA: 1s - loss: 10.6729Epoch 00008: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 10.6547 - val_loss: 20.1866
Epoch 9/50

11776/12188 [===========

Epoch 10/50

11776/12188 [===========================>..] - ETA: 1s - loss: 9.4934Epoch 00010: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 9.4870 - val_loss: 18.4566
Epoch 11/50

11776/12188 [===========================>..] - ETA: 1s - loss: 8.9356Epoch 00011: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 8.9303 - val_loss: 17.6210
Epoch 12/50

11776/12188 [===========================>..] - ETA: 1s - loss: 8.4319Epoch 00012: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 8.4229 - val_loss: 16.8048
Epoch 13/50

11776/12188 [===========================>..] - ETA: 1s - loss: 8.0381Epoch 00013: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 8.0088 - val_loss: 16.0121
Epoch 14/50

 7168/12188 [==

11776/12188 [===========================>..] - ETA: 1s - loss: 7.5444Epoch 00014: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 7.5233 - val_loss: 15.2803
Epoch 15/50

11776/12188 [===========================>..] - ETA: 1s - loss: 7.1654Epoch 00015: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 7.1368 - val_loss: 14.5362
Epoch 16/50

11776/12188 [===========================>..] - ETA: 1s - loss: 6.7199Epoch 00016: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 6.7342 - val_loss: 13.8420
Epoch 17/50

11776/12188 [===========================>..] - ETA: 1s - loss: 6.4404Epoch 00017: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 6.4248 - val_loss: 13.2086
Epoch 18/50

11776/12188 [===============

11776/12188 [===========================>..] - ETA: 1s - loss: 5.8241Epoch 00019: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 5.8287 - val_loss: 12.0862
Epoch 20/50

11776/12188 [===========================>..] - ETA: 1s - loss: 5.5850Epoch 00020: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 5.5932 - val_loss: 11.5003
Epoch 21/50

11776/12188 [===========================>..] - ETA: 1s - loss: 5.3370Epoch 00021: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 5.3265 - val_loss: 10.9933
Epoch 22/50

11776/12188 [===========================>..] - ETA: 1s - loss: 5.2081Epoch 00022: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 5.1947 - val_loss: 10.5884
Epoch 23/50

 8192/12188 [===============

11776/12188 [===========================>..] - ETA: 1s - loss: 4.9505Epoch 00023: saving model to checkpoints/2017-11-18_1136-seq2seq.hdf5

12188/12188 [==============================] - ETA: 0s - loss: 4.9367 - val_loss: 10.0716
Epoch 24/50

11776/12188 [===========================>..] - ETA: 1s - loss: 4.7661

# Validation

## Validation by feeding truth values

In [ ]:
def plot_preds(y_truth, y_pred):
    for junction in range(4):
        plt.figure
        plt.plot(y_truth[:,junction], color = 'green', label = 'Real traffic')
        plt.plot(y_pred[:,junction], color = 'red', label = 'Predicted traffic')
        plt.title('Traffic Forecasting at junction %i' % (junction+1))
        plt.xlabel('Number of hours from Start')
        plt.ylabel('Traffic')
        plt.legend()
        plt.show()

In [ ]:
y_pred = model.predict(X_test)

In [ ]:
plot_preds(y_test, y_pred)

In [ ]:
from sklearn.metrics import mean_squared_error
from math import sqrt

def rmse(y_true, y_pred):
    return sqrt(mean_squared_error(y_true, y_pred))

In [ ]:
rmse(y_test, y_pred)

## Validation by feeding predicted values

In [ ]:
def pred_seq(model, to_pred, w_size, num_preds=None):
    if num_preds == None:
        num_preds = to_pred.shape[0]
    current = to_pred[0]
    predicted = []
    for i in range(num_preds):
        predicted.append(np.round(model.predict(current[np.newaxis,:,:])[0,:]))
        current = current[1:]
        current = np.insert(current, [w_size-1], predicted[-1], axis=0)
    return np.asarray(predicted)

In [ ]:
seqpreds = pred_seq(model, X_test, seq_len)

In [ ]:
seqpreds.shape

In [ ]:
plot_preds(y_test, seqpreds)

In [ ]:
rmse(y_test, seqpreds)